connectors/memory/memory.go

```diff
@@ -245,9 +247,20 @@ func compareType(d1 dosa.FieldValue, d2 dosa.FieldValue) int8 {
 // Otherwise, search the partition for the exact same clustering keys. If there, fail
 // if not, then insert it at the right spot (sort.Search does most of the heavy lifting here)
 func (c *Connector) CreateIfNotExists(_ context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
-	return c.mergedInsert(ei.Def.Name, ei.Def.Key, values, func(into map[string]dosa.FieldValue, from map[string]dosa.FieldValue) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	err := c.mergedInsert(ei.Def.Name, ei.Def.Key, values, func(into map[string]dosa.FieldValue, from map[string]dosa.FieldValue) error {
 		return &dosa.ErrAlreadyExists{}
 	})
+	if err != nil {
+		return err
+	}
+	for iName, iDef := range ei.Def.Indexes {
+		// this error must be ignored, so we skip indexes when the value
+		// for one of the index fields is not specified
+		_ = c.mergedInsert(iName, ei.Def.UniqueKey(iDef.Key), values, overwriteValuesFunc)
```

Comment:
This makes me nervous. Right now, this relies on the fact that `mergedInsert` happens to only return an error when the value for one of the index fields isn't specified. In the future, if `mergedInsert` ever returns an error for another reason, we'll being ignoring that error here. It's also not super obvious that `mergedInsert` has this behavior, as evidenced by the fact that you had to document this strange ignoring of the error.

In other words, it feels to me like you're relying on an implementation detail of `mergeInsert` rather than just on it's contract.

I think the better way to do this is just not called merged insert if the values we're given don't have the index value in them. Doing this check
  - shouldn't be very hard to write
  - shouldn't need any new tests, since the behavior is the same
  - will make the code cleaner
  - and because the check will probably have some obvious name (like `hasRequiredIndexValue`) we can get rid of this docstring because the code will be self-documenting.


If you do go this route, the for loop will become a little more complex, but also be identical to the for-loop used down in upsert. You could probably extract the code into it's on function, `updateIndicies`.